keps/sig-apps/4026-job-creation-timestamp-annotation/README.md

```diff
@@ -0,0 +1,304 @@
+
+## Summary
+
+Currently, there is no supported way to get the original/expected initial scheduled timestamp for the job. This KEP proposes to set the the original scheduled time, `CreationTimestamp`, as an annotation in the job metadata.
+
+## Motivation
+
+### Goals
+
+- Set job creation timestamp as an annotation on the job.
+- Adding the annotation should not be disruptive to existing workloads.
+
+### Non-Goals
+
+## Proposal
+
+At a high level, the proposal is to modify the CronJob controller to set the job creation timestamp as a job annotation. The details of this are outlined in the Design Details section below.
+
+Job creation timestamp annotation: `batch.kubernetes.io/job-creation-timestamp`
+
+### User Stories (Optional)
+
+#### Story 1
+
+As a user, I would like to get the job's creation timestamp that this job was expected to be running.
+
+#### Story 2
+
+### Notes/Constraints/Caveats (Optional)
+
+### Risks and Mitigations
+
+One thing that must be considered is how enabling this new feature will interact with existing workloads. There are a couple of options:
+
+1. Only inject the job annotation for *newly created CronJobs*. We can track this by annotating newly created CronJobs to distinguish existing ones from newly created ones.
+Using this strategy, either none of the jobs have this annotation, or all of them do, which will provide a more consistent user experience. However, in the case of a cluster downgrade to a version without this feature, new jobs would start getting created without this annotation again.
+
+1. Inject the annotation on *all jobs* (jobs existing prior to feature enablement and jobs created after feature enablement). However, retroactively modifying jobs of existing workloads would risk being too disruptive to existing workloads which may have logic depending on job annotations, so this option should not be considered.
+
+Both options 1 and 2 will not be disruptive to existing workloads. Option 1 is more straightforward and does not risk locking us into adding this somewhat
+hacky annotation to Jobs indefinitely like Option 2 does. On the other hand, outside of the cluster downgrade edge case, Option 2 will
+ensure consistency within a single CronJob and therefore a more predictable user experience.
+
+After considering these trade-offs, we propose to move forward with Option 1 for simplicity and to avoid being stuck adding this annotation to Jobs. In addition, the downside of existing workloads having only a subset of jobs with the new annotation will not cause any serious issues.
+
+## Design Details
+
+The CronJob controller will only need a minor update to the [getJobFromTemplate2](https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/cronjob/utils.go#L188) function, to add the job creation timestamp as the job annotation `batch.kubernetes.io/job-creation-timestamp`. The creation timestamp is represented in `RFC3339` form and is in `UTC`.
```

Comment:
```suggestion
The CronJob controller will only need a minor update to the [getJobFromTemplate2](https://github.com/kubernetes/kubernetes/blob/6f2cd1b5bd76e336c08846281db1aacca1e7dd13/pkg/controller/cronjob/utils.go#L188) function, to add the job creation timestamp as the job annotation `batch.kubernetes.io/job-creation-timestamp`. The creation timestamp is represented in `RFC3339` form and is in `UTC`.
```

I saw in another KEP this recommendation to not include code links from master as those don't age well.

Reply:
Updated.